### PR TITLE
Editing of custom fields of type "Date"

### PR DESCRIPTION
**Pull request by knight-bubble**
[`* `#19614` Date custom fields are not editable`](http://community.openproject.org/work_packages/19614)
also
 [`* `#17219`  Inplace editing for custom values of format "Date"`](http://community.openproject.org/work_packages/17219)

### CODE REVIEW

**Comment by TeatroIO** [comment]
I've prepared a stage to preview changes. <a href='http://hirtie-maxim-feature-17219-custom-values-date.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://hirtie-maxim-feature-17219-custom-values-date.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

**Comment by NobodysNightmare** [comment]
@hirtie-maxim You need to rebase or cherry-pick your commits against `release/4.1`:

e.g.

```
git fetch upstream
git reset --hard upstream/release/4.1
git cherry-pick 30bb13eb322407db021d9854ec9822c8c3a669b8
git cherry-pick 9c7322a2e0f6d270e0dffa1c85ec664acaead8ae
```

Optionally the PR should be re-issued against `release/4.1`, however, this is optional as long as the reviewer is careful when merging your PR.

P.S.: Looking at the network graph you have at least two other PRs, that need to be rebased against `release/4.1`.

**Comment by NobodysNightmare** [comment]
For empty (no value) date custom fields I found two issues:
1. The text should read `-` (like for any other empty CF)
2. Unless focussed the text field is not high enough (probably because there is no text in it)

small-and-empty

**Comment by NobodysNightmare** [comment]
OMG! What are you doing with dates?

When I set the eleventh day of July 2010 as date:

You are sending `2010-11-07` via API (the seventh day of November 2010).
You are displaying the date as `11/07/2010` to the user.

When I set the thirteenth day of July 2010 as date:

You are sending `2010-07-13` via API (that is the correct date)
You are displaying the date as `07/13/2010` to the user (another order than used before).

It is really frightening, that the date representation flips the order of its components as soon as the date surpasses the **12**.

That also leads to the calendar widget opening in November 2020, whenever a date between the 13th and the 31st of a month is selected.

## For the record

The ISO date (sent via API) is **always** `YYYY-MM-DD`, that is the date component is always last, the month always second.

The date displayed to the user should **as the very least** be **_consistent**_. There is actually a instance wide setting on the display format used for dates. Ideally one would use this format for user interaction.

**Comment by NobodysNightmare** [comment]
Oh lord... this issue is also real for the already merged stuff -.-

**Comment by NobodysNightmare** [comment]
title-text

Apparently the title text does not update after saving a date...

**Comment by myabc** [comment]
@NobodysNightmare @hirtie-maxim I had some comments about the code itself, but was saving them until we make a new PR against `release/4.1`.

File: frontend/tests/integration/specs/work-packages/work-packages-modal-focus-spec.js [hunk 1]
```diff
@@ -57,7 +57,7 @@ describe('OpenProject', function() {
     });
   });
 
-  it('grouping modal should focus', function() {
+  xit('grouping modal should focus', function() {
```
Comment (NobodysNightmare) [inline frontend/tests/integration/specs/work-packages/work-packages-modal-focus-spec.js:60]: I don't see why you should disable those three specs in your PR.

File: frontend/app/work_packages/directives/inplace_editor/custom/editable/inplace-editor-date-directive.js [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+          addClearButton(input);
+        },
+        onChangeMonthYear: function() {
+          addClearButton(input);
+        },
+        onClose: function(selectedDate) {
+          if(!selectedDate || selectedDate === '' || selectedDate === prevDate) {
+            return;
+          }
+          var parsedDate = parseDate(selectedDate, 'DD/MM/YYYY');
```
Comment (NobodysNightmare) [inline frontend/app/work_packages/directives/inplace_editor/custom/editable/inplace-editor-date-directive.js]: I've spotted the `DD/MM/YYYY` constant three times in this file.

Leading to two questions:
1. Why is it not a common constant used in all three places?
2. Why is it a constant at all and not defined by the `ConfigurationService.dateFormat`?

**Comment by floriank** [comment]
please reissue this against `release/4.1` and fix the remaining bugs @hirtie-maxim

**Comment by floriank** [comment]
Closed as this is superseded by #2904

**Pull request closed without merging**